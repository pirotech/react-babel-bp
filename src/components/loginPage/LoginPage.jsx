import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import UiLanguageSelector from '../../shared/components/UiLanguageSelector/UiLanguageSelector';
import authActions from '../../store/actions/authActions';
import authSelector from '../../store/selectors/authSelector';
import './LoginPage.scss';

class LoginPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      password: ''
    };
  }

  componentDidMount() {
    const {history, isAuthorized} = this.props;
    if (isAuthorized) {
      history.push('/');
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const {history, isAuthorized} = this.props;
    if (isAuthorized) {
      history.push('/');
    }
  }

  onFieldChange = (name, value) => {
    this.setState({
      [name]: value
    });
  };

  login = () => {
    const {login} = this.props;
    const {username, password} = this.state;
    const preparedUsername = username.trim();
    const preparedPassword = password.trim();
    if (!preparedUsername || !preparedPassword) {
      return;
    }

    login(preparedUsername, preparedPassword);
  };

  render() {
    const {t} = this.props;

    return (
      <div className="login-page">
        <UiLanguageSelector/>
        <p>{t('username')}</p>
        <input
          type="text"
          onChange={event => this.onFieldChange('username', event.target.value)}
        />
        <p>{t('password')}</p>
        <input
          type="text"
          onChange={event => this.onFieldChange('password', event.target.value)}
        />
        <button
          className="login-page__submit-button"
          onClick={this.login}
        >{t('loginButton')}</button>
      </div>
    );
  }
}

LoginPage.propTypes = {};

const mapStateToProps = (state) => ({
  isAuthorized: authSelector.getAuthorized(state)
});
const mapDispatchToProps = (dispatch) => (
  bindActionCreators(authActions, dispatch)
);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation('loginPage')(LoginPage));