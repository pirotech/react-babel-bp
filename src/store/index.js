import { createBrowserHistory } from 'history';
import { applyMiddleware, createStore, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import createRootReducer from './reducers';
import authSaga from './sagas';
import {routerMiddleware} from 'connected-react-router';
import logger from 'redux-logger';

export const history = createBrowserHistory();

const sagaMiddleWare = createSagaMiddleware();

export const store = createStore(
  createRootReducer(history),
  {},
  compose(applyMiddleware(
    routerMiddleware(history),
    sagaMiddleWare,
    logger
  ))
);

sagaMiddleWare.run(authSaga);
