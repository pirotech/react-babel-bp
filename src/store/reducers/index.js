import { connectRouter } from 'connected-react-router';
import { combineReducers } from 'redux';
import authReducer from './authReducer';

const createRootReducer = (history) => {
  return combineReducers({
    router: connectRouter(history),
    auth: authReducer
  });
};

export default createRootReducer;