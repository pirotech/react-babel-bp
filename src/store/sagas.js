import { takeEvery, put } from 'redux-saga/effects';
import { usersService } from '../services';
import { LOGIN_REQUESTED, LOGIN_SUCCESS, LOGIN_FAILURE } from './actionTypes';

function* loginRequested({payload}) {
  try {
    const {username, password} = payload;
    const [promise] = usersService.login(username, password);
    const {token} = yield promise;
    yield put({
      type: LOGIN_SUCCESS,
      payload: {
        token
      }
    });
  } catch (e) {
    yield put({
      type: LOGIN_FAILURE
    });
  }
}

function* authSaga() {
  yield takeEvery(LOGIN_REQUESTED, loginRequested);
}

export default authSaga;