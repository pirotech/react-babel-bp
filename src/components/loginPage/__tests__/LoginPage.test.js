import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import LoginPage from '../LoginPage';
import { LOGIN_REQUESTED, LOGIN_SUCCESS } from '../../../store/actionTypes';
import { history, store } from './__mocks__';

jest.mock('react-i18next', () => ({
  withTranslation: () => Component => {
    Component.defaultProps = {
      ...Component.defaultProps,
      t: key => key,
      i18n: {
        language: 'en',
        changeLanguage: (language) => {
        }
      }
    };
    return Component;
  }
}));

describe('Login page', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store.emptyToken}>
        <LoginPage history={history}/>
      </Provider>
    );
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should not send empty form', () => {
    // arrange
    const loginPage = wrapper.find(LoginPage).children();
    const button = wrapper.find('.login-page__submit-button');

    // act
    button.simulate('click');

    // assert
    const expectedAction = {
      type: LOGIN_REQUESTED,
      payload: {
        username: 'admin',
        password: 'admin'
      }
    };
    expect(store.emptyToken.getActions()).not.toContainEqual(expectedAction);
  });

  it('should send with form with content', () => {
    // arrange
    const loginPage = wrapper.find(LoginPage).children();
    const button = wrapper.find('.login-page__submit-button').first();
    loginPage.setState({
      username: 'admin',
      password: 'admin'
    });
    store.emptyToken.dispatch({
      type: LOGIN_SUCCESS,
      dispatch: {
        token: '12345'
      }
    });
    console.log(store.emptyToken);

    // act
    button.simulate('click');

    // assert
    const expectedAction = {
      type: LOGIN_REQUESTED,
      payload: {
        username: 'admin',
        password: 'admin'
      }
    };
    expect(store.emptyToken.getActions()).toContainEqual(expectedAction);
  });

  it('should redirect with valid token', () => {
    // arrange
    wrapper = mount(
      <Provider store={store.token}>
        <LoginPage history={history}/>
      </Provider>
    );

    // act

    // assert
    expect(history.push).toHaveBeenCalled();
  });
});
