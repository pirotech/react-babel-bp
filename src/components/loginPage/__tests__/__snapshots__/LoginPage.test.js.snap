// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page should redirect with valid token 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(LoginPage)
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      }
    }
  >
    <LoginPage
      history={
        Object {
          "push": [MockFunction] {
            "calls": Array [
              Array [
                "/",
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        }
      }
      i18n={
        Object {
          "changeLanguage": [Function],
          "language": "en",
        }
      }
      isAuthorized={true}
      login={[Function]}
      t={[Function]}
    >
      <div
        className="login-page"
      >
        <UiLanguageSelector
          i18n={
            Object {
              "changeLanguage": [Function],
              "language": "en",
            }
          }
          t={[Function]}
        >
          <div
            className="ui-language-selector"
          >
            <button
              onClick={[Function]}
            >
              en
            </button>
          </div>
        </UiLanguageSelector>
        <p>
          username
        </p>
        <input
          onChange={[Function]}
          type="text"
        />
        <p>
          password
        </p>
        <input
          onChange={[Function]}
          type="text"
        />
        <button
          className="login-page__submit-button"
          onClick={[Function]}
        >
          loginButton
        </button>
      </div>
    </LoginPage>
  </Connect(LoginPage)>
</Provider>
`;

exports[`Login page should render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(LoginPage)
    history={
      Object {
        "push": [MockFunction],
      }
    }
  >
    <LoginPage
      history={
        Object {
          "push": [MockFunction],
        }
      }
      i18n={
        Object {
          "changeLanguage": [Function],
          "language": "en",
        }
      }
      isAuthorized={false}
      login={[Function]}
      t={[Function]}
    >
      <div
        className="login-page"
      >
        <UiLanguageSelector
          i18n={
            Object {
              "changeLanguage": [Function],
              "language": "en",
            }
          }
          t={[Function]}
        >
          <div
            className="ui-language-selector"
          >
            <button
              onClick={[Function]}
            >
              en
            </button>
          </div>
        </UiLanguageSelector>
        <p>
          username
        </p>
        <input
          onChange={[Function]}
          type="text"
        />
        <p>
          password
        </p>
        <input
          onChange={[Function]}
          type="text"
        />
        <button
          className="login-page__submit-button"
          onClick={[Function]}
        >
          loginButton
        </button>
      </div>
    </LoginPage>
  </Connect(LoginPage)>
</Provider>
`;
